monai/apps/utils.py

```diff
@@ -54,20 +54,25 @@ def get_logger(
     """
     Get a `module_name` logger with the specified format and date format.
     By default, the logger will print to `stdout` at the INFO level.
-    If `module_name` is `None`, return the root logger.
-    `fmt` and `datafmt` are passed to a `logging.Formatter` object
+    - If `module_name` is None, the root logger is returned.
+    - `fmt` and `datefmt` are passed to `logging.Formatter` for formatting.
     (https://docs.python.org/3/library/logging.html#formatter-objects).
-    `logger_handler` can be used to add an additional handler.
+    - If `logger_handler` is provided, it will be added to the logger.
     """
-    adds_stdout_handler = module_name is not None and module_name not in logging.root.manager.loggerDict
+    # Retrieve or create a logger for the module
     logger = logging.getLogger(module_name)
-    logger.propagate = False
     logger.setLevel(logging.INFO)
-    if adds_stdout_handler:  # don't add multiple stdout or add to the root
-        handler = logging.StreamHandler(sys.stdout)
-        formatter = logging.Formatter(fmt=fmt, datefmt=datefmt)
-        handler.setFormatter(formatter)
-        logger.addHandler(handler)
+
+    # Check if stdout handler should be added (avoid adding multiple times)
+    if module_name and module_name not in logging.root.manager.loggerDict:
```

Comment:
Hi @ericspod ,

I am not very familiar with these python libs, could you please help double confirm it?

Thanks in advance.

Reply:
This is different from the current code in that a `StreamHandler` is never created in this function. Previously, if the function is called and there wasn't a handler for the given module name, this would be created, but here we only try to find that handler and then set it's formatter if found. No code is now automatically creating the `StreamHandler` instance writing to stdout. I don't see how this solves the problem of there being things logged to stdout and not files (or vice versa). I have noticed this discrepency in the past and felt at some point we should have a logging module to keep better track of these sorts of things, and have a single logger rather than one per module which seems needless.

Reply:
On another note, there's a number of things in this file that should go into their own places in `monai.utils`, such as moving the download functions to a `monai/utils/download.py` file. I was working some importation to clean up how modules are loaded but this broke some bundles, but I want to return to that to clean things up and reduce this file in size.

Reply:
Hi @ericspod, Let me briefly explain the changes I've made. When we use `logging.getLogger(module_name)` to create a logger, it gets registered under ``logging.root. Previously, we would check if the logger wasn't already in the root and, if not, add a `StreamHandler` while also formatting the output.

However, I noticed that we set `propagate = False`, which means the existing `StreamHandler` in the root won't be used. Now, I’ve adjusted the code by not setting this argument, which allows us to utilize the root's existing `StreamHandler` and only update the format if needed.

This change helps resolve the issue because in the bundle, we configure root as the logger. If we set `propagate = False`, logging won’t happen automatically, and we would need to manually configure loggers for each module (e.g., monai.apps).
https://github.com/Project-MONAI/model-zoo/blob/dev/models/brats_mri_segmentation/configs/logging.conf#L2

Hope I have clarified the change here.

Reply:
Hi @KumoLiu I get what you're wanting to do now thanks, however my other point on the logic being quite different stands in that there isn't a handler being created at the same times now as was the case in the previous version of this code. The conditions have been moved around and this effects the final logic, so I'd double check through tracing or some other way that the behaviour is the same.

Reply:
Hi @ericspod, I reverted the changes in get_logger after discovering that the issue with StatsHandler not logging through the FileHandler was due to the logger name not reflecting the correct hierarchy. I've now updated the logger name to monai.xxxx.stats_handlers, and it's being logged correctly.

please help review it again, thanks.

Reply:
Ok I see the changes that were done and I'm good with them, but did you not also want to remove the `propagate = False` part from the `get_logger` function?

Reply:
Yes, we don't need to remove that since for now all the logging logic can be handled by "monai" this root logger. So what we need to add in the logging.conf is some thing like:
Add a new logger named monai and also apply the filehandler to it, and set `propagate=0` in it.
```
[loggers]
keys=root, monai

[logger_monai]
level=INFO
handlers=consoleHandler, fileHandler
qualname=monai
propagate=0
```

Reply:
If I remove that from the `get_logger`, the logging output will be **duplicated** by default, since MONAI also adds a `StreamHandler`.

Reply:
Sounds good, thanks!